### PR TITLE
[Merged by Bors] - Add Beziers to `bevy_math`

### PR DESCRIPTION
**Pull request by aevyrie**
# Objective

- Adds foundational math for Bezier curves, useful for UI/2D/3D animation and smooth paths.

https://user-images.githubusercontent.com/2632925/218883143-e138f994-1795-40da-8c59-21d779666991.mp4

## Solution

- Adds the generic `Bezier` type, and a `Point` trait. The `Point` trait allows us to use control points of any dimension, as long as they support vector math. I've implemented it for `f32`(1D), `Vec2`(2D), and `Vec3`/`Vec3A`(3D).
- Adds `CubicBezierEasing` on top of `Bezier` with the addition of an implementation of cubic Bezier easing, which is a foundational tool for UI animation.
  - This involves solving for $t$ in the parametric Bezier function $B(t)$ using the Newton-Raphson method to find a value with error $\leq$ 1e-7, capped at 8 iterations.
- Added type aliases for common Bezier curves: `CubicBezier2d`, `CubicBezier3d`, `QuadraticBezier2d`, and `QuadraticBezier3d`. These types use `Vec3A` to represent control points, as this was found to have an 80-90% speedup over using `Vec3`.
- Benchmarking shows quadratic/cubic Bezier evaluations $B(t)$ take \~1.8/2.4ns respectively. Easing, which requires an iterative solve takes \~50ns for cubic Beziers. 

---

## Changelog

- Added `CubicBezier2d`, `CubicBezier3d`, `QuadraticBezier2d`, and `QuadraticBezier3d` types with methods for sampling position, velocity, and acceleration. The generic `Bezier` type is also available, and generic over any degree of Bezier curve.
- Added `CubicBezierEasing`, with additional methods to allow for smooth easing animations.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Very useful as a primitive. I'm not sure how I feel about a trait specifically for *cubic* beziers; a const generic for the number of control points feels more correct.

That's 110% fine to punt to later work though.

**Comment by aevyrie** [comment]
> Very useful as a primitive. I'm not sure how I feel about a trait specifically for _cubic_ beziers; a const generic for the number of control points feels more correct.
> 
> That's 110% fine to punt to later work though.

Cubic Beziers always have 4 control points. An algorithm for higher-order Bezier curves is not included here, and could become a performance footgun. If users need more control points per spline, they should (probably) either chain cubic Beziers, or use a different type of spline. 

I created a trait because the math is the same for cubic Beziers of any dimension. I originally had a generic function, but wanted to be able to use `self`, so I made a trait. That had the downside of requiring the trait to be imported, so I "re-exported" the trait method by adding delegates to the concrete types. Because of this, I could go back to using a generic function while still being able to use `self` via the delegate.

**Comment by alice-i-cecile** [comment]
That's a good explanation: I'm happy with this approach.

File: crates/bevy_math/src/bezier.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    /// Returns the point along the Bezier curve at the supplied parametric value `t`.
+    pub fn evaluate_at(&self, t: f32) -> Vec2 {
+        CubicBezier::evaluate_at(self, t)
+    }
+
+    /// Iterate over points in the bezier curve from `t = 0` to `t = 1`.
+    pub fn to_points(&self, subdivisions: i32) -> Vec<Vec2> {
+        CubicBezier::to_points(self, subdivisions)
+    }
+}
```
Comment (mockersf) [inline crates/bevy_math/src/bezier.rs:55]: why do you need to reimplement those from the trait?

File: crates/bevy_math/src/bezier.rs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        x: CubicBezier1d,
+        y: CubicBezier1d,
+    }
+
+    impl CubicBezierEasing {
+        /// Construct a cubic bezier curve for animation easing, with control points P1 and P2.
+        ///
+        /// This is equivalent to the syntax used to define cubic bezier easing functions in, say, CSS:
+        /// `ease-in-out = cubic-bezier(0.42, 0.0, 0.58, 1.0)`.
+        pub fn new(p1_x: f32, p1_y: f32, p2_x: f32, p2_y: f32) -> Self {
```
Comment (mockersf) [inline crates/bevy_math/src/bezier.rs:110]: should `p1` and `p2` be `Vec2`?

File: crates/bevy_math/src/bezier.rs (hunk 1 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:55]: I added these delegates so users of the Bezier types don't need to import the trait, which I find to be pretty annoying as a user. The trait is really an implementation detail to reduce code duplication, and doesn't need to be used publicly.

File: crates/bevy_math/src/bezier.rs (hunk 2 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:110]: In this case, no, I don't think so. Cubic Bezier easing is [commonly used](https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function) with the syntax `cubic-bezier(0.42, 0.0, 1.0, 1.0)`. I've mirrored that syntax here as users will probably expect the constructor to look the same.

Comment (mockersf) [reply crates/bevy_math/src/bezier.rs:110]: If this were in `bevy_ui` I would agree, but not in `bevy_math`. CSS has one of the crappiest type system in the world in my opinion 😄

File: crates/bevy_math/src/bezier.rs (hunk 1 above)
Comment (mockersf) [reply crates/bevy_math/src/bezier.rs:55]: Not a fan, but why not. Then:
* could you mark them as `inline`?
* could you make sure the docs are the same between all imply and the trait?

File: crates/bevy_math/src/bezier.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:110]: I agree with Francois: more types please...

Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:110]: > CSS has one of the crappiest type system in the world in my opinion 😄

FWIW, this is not just CSS, in fact I linked to the Android/Kotlin docs. This is also true for [flutter](https://api.flutter.dev/flutter/animation/Cubic/Cubic.html), [Qt](https://doc.qt.io/qt-6/qml-qtquick-pathcubic.html), [Elm](https://github.com/Dandandan/Easing/blob/master/Easing.elm#L161), and everywhere else I've looked. From what I can tell, this is basically standard, and would be a bit odd for a type designed for cubic Bezier animation not to have a constructor that takes an `f32` 4-tuple. I've also done a refactor pass based on other feedback that has resulted in the `CubicBezierEasing` type simply being a pair of `Vec2` control points.

Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:110]: I found an implementation that is strongly typed, from Apple: https://developer.apple.com/documentation/uikit/uicubictimingparameters/1649854-init

**Comment by aevyrie** [comment]
We need some tests to verify this is working as expected. Any suggestions?

1. Draw an animation curve that exactly represents a diagonal line, and compare samples to an explicit function:
$y = 0.5 * x$ = `Bezier.evaluate_y_at(t)`
2. Use another program to compute the values, and hardcode those as correct answers in the test

**Comment by aevyrie** [comment]
Huh, looks like there is a duplicate dep, though I haven't touched the lockfile.

File: crates/bevy_math/src/bezier.rs [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    pub fn sample(&self, subdivisions: i32, sample_function: fn(&Self, f32) -> P) -> Vec<P> {
+        (0..=subdivisions)
+            .map(|i| {
+                let t = i as f32 / subdivisions as f32;
+                sample_function(self, t)
+            })
+            .collect()
+    }
+
+    /// Split the Bezier curve into `subdivisions` across the length of the curve from t = `0..=1`.
```
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:81]: Ditto on the methods below.

File: crates/bevy_math/src/bezier.rs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    /// Given a `time` within `0..=1`, remaps to a new value using the cubic Bezier curve as a
+    /// shaping function, for which when plotted `x = time` and `y = animation progress`. This will
+    /// return `0` when `t = 0`, and `1` when `t = 1`.
+    pub fn ease(&self, time: f32) -> f32 {
+        let x = time.clamp(0.0, 1.0);
+        let t = self.find_t_given_x(x);
+        self.evaluate_y_at(t)
+    }
+
+    /// Compute the x-coordinate of the point along the Bezier curve at `t`.
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:134]: It's not clear from these docs / methods how to get the xy pair.

File: crates/bevy_math/src/bezier.rs [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    #[inline]
+    pub fn dx_dt(&self, t: f32) -> f32 {
+        bezier_impl::velocity([0.0, self.p1.x, self.p2.x, 1.0], t)
+    }
+
+    /// Solve for the parametric value `t` that corresponds to the given value of `x`.
+    #[inline]
+    pub fn find_t_given_x(&self, x: f32) -> f32 {
+        let mut t_guess = x;
+        (0..Self::MAX_ITERS).any(|_| {
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:156]: This should probably note in a comment that we're using Newton's method.

File: crates/bevy_math/src/bezier.rs [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        t_guess.clamp(0.0, 1.0)
+    }
+}
+
+/// Generic implementations for sampling cubic Bezier curves. Consider using the methods on
+/// [`Bezier`] for more ergonomic use.
+pub mod bezier_impl {
+    use super::Point;
+
+    /// Compute the bernstein basis polynomial for iteration `i`, for a Bezier curve with with
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:176]:
```suggestion
    /// Compute the Bernstein basis polynomial for iteration `i`, for a Bezier curve with with
```

File: crates/bevy_math/src/bezier.rs [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    /// degree `n`, at `t`.
+    #[inline]
+    pub fn bernstein_basis(n: usize, i: usize, t: f32) -> f32 {
+        (1. - t).powi((n - i) as i32) * t.powi(i as i32)
+    }
+
+    /// Efficiently compute the binomial coefficient
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:183]:
```suggestion
    /// Efficiently compute the binomial coefficient of `n` choose `k`. 
```

File: crates/bevy_math/src/bezier.rs [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+            .sum()
+    }
+
+    /// Compute the second derivative B''(t) of Bezier curve B(t) of degree `N-1` at the given
+    /// parametric value `t` with respect to t.
+    #[inline]
+    pub fn acceleration<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        debug_assert!(
+            N > 2,
+            "Acceleration can only be computed on Bezier curves with a degree of 2 or higher"
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:236]: Shouldn't this always return 0 for lower dimensional curves instead?

File: crates/bevy_math/src/bezier.rs [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+            .map(|i| p[i] * binomial_coeff(degree, i) as f32 * bernstein_basis(degree, i, t))
+            .sum()
+    }
+
+    /// Compute the first derivative B'(t) of Bezier curve B(t) of degree `N-1` at the given
+    /// parametric value `t` with respect to t.
+    #[inline]
+    pub fn velocity<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        debug_assert!(
+            N > 1,
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:215]: Shouldn't this always return 0 for lower dimensional curves instead?

File: crates/bevy_math/src/bezier.rs (hunk 7 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:183]: Ah, is that how you say it in plain English? This was my first time running into binomial coefficient notation, and initially I had no idea I was supposed to evaluate it down to a value.

File: crates/bevy_math/src/bezier.rs (hunk 4 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:134]: I don't understand your request. I thought the doc comments on `ease` explained what was happening. Although perhaps most of these methods should not be public to avoid confusion? Users probably only ever need to use `bezier.ease(t)`.

File: crates/bevy_math/src/bezier.rs (hunk 9 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:215]: Hmmm, that's a good point. We can rely on the `Default` impl returning 0 for all our `Point` types.

File: crates/bevy_math/src/bezier.rs (hunk 2 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs]: I think this is close enough to be the best of both worlds:

```rs
pub fn new(p1: impl Into<Vec2>, p2: impl Into<Vec2>)
```

Which means you can use it with minimal boilerplate:
```rs
let cubic_bezier = CubicBezierEasing::new((0.8, 0.0), (0.2, 1.0));
let t = cubic_bezier.ease(time);
```

Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs]: Good, I like that.

**Comment by aevyrie** [comment]
CI is failing on: https://github.com/bevyengine/bevy/issues/7654

File: crates/bevy_math/src/bezier.rs [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+    const MAX_ERROR: f32 = 1e-7;
+
+    /// Maximum number of iterations during bezier solve
+    const MAX_ITERS: u8 = 8;
+
+    /// Given a `time` within `0..=1`, returns an eased value within `0..=1` that follows the cubic
+    /// Bezier curve instead of a straight line.
+    ///
+    /// The start and endpoints will match: ease(0) = 0 and ease(1) = 1.
+    pub fn ease(&self, time: f32) -> f32 {
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:137]: This uses `time`, while the others use `t`. Mild preference for `t`, but I only actually care about consistency.

File: crates/bevy_math/src/bezier.rs (hunk 4 above)
Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:143]: Fundamentally, it's not clear what `t` is here in context. Is it the distance along the path?

Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:143]: With these private, I'm fine without a `evaluate_at_t(&self, t: f32) -> Vec2` method, which is what these existing methods implied.

File: crates/bevy_math/src/bezier.rs [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+    /// Maximum allowable error for iterative bezier solve
+    const MAX_ERROR: f32 = 1e-7;
+
+    /// Maximum number of iterations during bezier solve
+    const MAX_ITERS: u8 = 8;
+
+    /// Given a `time` within `0..=1`, returns an eased value within `0..=1` that follows the cubic
+    /// Bezier curve instead of a straight line.
+    ///
+    /// The start and endpoints will match: ease(0) = 0 and ease(1) = 1.
```
Comment (alice-i-cecile) [inline crates/bevy_math/src/bezier.rs:136]: Doc tests for these invariants would be a nice touch.

File: crates/bevy_math/src/bezier.rs (hunk 10 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:137]: No. Time is the x-axis of the Bezier. The reason easing is complex is that you need to take this `x` value, solve for the corresponding `t`, then compute `y`.

File: crates/bevy_math/src/bezier.rs [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+/// A cubic Bezier curve in 3D space
+pub type CubicBezier3d = Bezier<Vec3A, 4>;
+/// A quadratic Bezier curve in 2D space
+pub type QuadraticBezier2d = Bezier<Vec2, 3>;
+/// A quadratic Bezier curve in 3D space
+pub type QuadraticBezier3d = Bezier<Vec3A, 3>;
+
+/// A Bezier curve with `N` control points, and dimension defined by `P`.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Bezier<P: Point, const N: usize>(pub [P; N]);
```
[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:74]: Take a look now, I think I've resolved this.

File: crates/bevy_math/src/bezier.rs (hunk 10 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:234]: Added docs

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_math/src/bezier.rs [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+/// to ensure the animation begins at 0, and ends at 1.
+#[derive(Default, Clone, Copy, Debug, PartialEq)]
+pub struct CubicBezierEasing {
+    /// Control point P1 of the 2D cubic Bezier curve. Controls the start of the animation.
+    pub p1: Vec2,
+    /// Control point P2 of the 2D cubic Bezier curve. Controls the end of the animation.
+    pub p2: Vec2,
+}
+
+impl CubicBezierEasing {
```
[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:164]: I considered this, but wasn't sure if I should add these - they would certainly be nice to have.

[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:164]: Also, is there a license for this? Can we copy it feely even if unlicensed? Is it even possible to "own" IP of a Bezier curve?

[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:164]: Root repo is GPLv3. The collection and ordering is almost certainly copyrightable; the exact values are not.

I'm against including them from this source.

[the remark this answers is not in the document]
Comment (doup) [reply crates/bevy_math/src/bezier.rs:164]: Ops, sorry, I didn't even think on licensing… 😅 

Apparently the source is "Robert Penner" and an old Macromedia Flash MX book (before Adobe!), the related page is here: http://robertpenner.com/easing/ The easing functions are licensed under BSD/MIT… but I don't know where the cubic-bezier approximations come from. It should be both Apache/MIT, or it's OK if it's only MIT?

It should be fairly easy (although, time consuming) to create our own approximations by drawing on top the curves. I suppose that should be fine; but the names could be problematic?

[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:164]: The W3C CSS spec has ease-in/out/etc listed, might be able to use that. Material design also has similar, and is open source.

[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply crates/bevy_math/src/bezier.rs:164]: Only MIT is fine IMO; but we'll want a little license note in the docs or comments for the file.

[the remark this answers is not in the document]
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:164]: I think we should punt this to another PR, the license discussion will be distracting, and is not necessary for this ticket.
Definitely a valuable addition though.

[the remark this answers is not in the document]
Comment (doup) [reply crates/bevy_math/src/bezier.rs:164]: Totally, let's decide this later on. 👍

**Comment by aevyrie** [comment]
bors try

**Comment by aevyrie** [comment]
bors try

**Comment by aevyrie** [comment]
bors try

File: crates/bevy_math/src/bezier.rs [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+        let k = match k > n - k {
+            true => n - k,
+            false => k,
+        };
+        while i < k {
+            result *= n - i;
+            result /= i + 1;
+            i += 1;
+        }
+        result
```
Comment (JoJoJet) [inline crates/bevy_math/src/bezier.rs:329]:
```suggestion
        let k = usize::min(k, n - k);
        (0..k).fold(1, |val, i| val * (n - i) / (i + 1))
```

**Review by JoJoJet — COMMENTED** [review]
Seems like it will be very useful to have this standardized. I've needed bezier curves in my own projects before, it will be nice to have a foundation in bevy itself.

File: crates/bevy_math/src/bezier.rs [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+    + Debug
+    + Clone
+    + PartialEq
+    + Copy
+{
+}
+impl Point for Vec3 {} // 3D
+impl Point for Vec3A {} // 3D
+impl Point for Vec2 {} // 2D
+impl Point for f32 {} // 1D
```
Comment (james7132) [inline crates/bevy_math/src/bezier.rs:25]: Not necessarily blocking, but it would be interesting to see if we should consider how this would interact with the `Animatable` trait in #4482.

File: crates/bevy_math/src/bezier.rs [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+    /// Compute the second derivative B''(t) of this bezier at `t` with respect to t.This is the
+    /// instantaneous acceleration of a point tracing the Bezier curve from t = 0 to 1.
+    pub fn acceleration(&self, t: f32) -> P {
+        generic::acceleration(self.0, t)
+    }
+
+    /// Split the Bezier curve of degree `N-1` into `subdivisions` evenly spaced `t` values across
+    /// the length of the curve from t = `0..=1`, and sample with the supplied `sample_function`.
+    #[inline]
+    pub fn sample(&self, subdivisions: i32, sample_function: fn(&Self, f32) -> P) -> Vec<P> {
```
Comment (james7132) [inline crates/bevy_math/src/bezier.rs:123]: Might be better to return an iterator here instead of a `Vec` to allow users of the API to decide whether or not to allocate.

File: crates/bevy_math/src/bezier.rs [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+    }
+}
+
+/// Generic implementations for sampling Bezier curves. Consider using the methods on [`Bezier`] for
+/// more ergonomic use.
+pub mod generic {
+    use super::Point;
+
+    /// Compute the Bernstein basis polynomial for iteration `i`, for a Bezier curve with with
+    /// degree `n`, at `t`.
```
Comment (james7132) [inline crates/bevy_math/src/bezier.rs:309]:
```suggestion
    /// degree `n`, at `t`.
    ///
    /// For more information, see https://en.wikipedia.org/wiki/Bernstein_polynomial.
```

File: crates/bevy_math/src/bezier.rs [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+        t_guess.clamp(0.0, 1.0)
+    }
+}
+
+/// Generic implementations for sampling Bezier curves. Consider using the methods on [`Bezier`] for
+/// more ergonomic use.
+pub mod generic {
+    use super::Point;
+
+    /// Compute the Bernstein basis polynomial for iteration `i`, for a Bezier curve with with
```
Comment (james7132) [inline crates/bevy_math/src/bezier.rs:308]:
```suggestion
    /// Compute the Bernstein basis polynomial for iteration `i`, for a Bezier curve with
```

File: crates/bevy_math/src/bezier.rs [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+    #[inline]
+    pub fn acceleration<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        if N <= 2 {
+            return P::default(); // Zero for numeric types
+        }
+        let p = control_points;
+        let degree = N - 1;
+        let degree_vel = N - 2; // the velocity Bezier is one degree lower than the position Bezier
+        let degree_accel = N - 3; // the accel Bezier is one degree lower than the velocity Bezier
+        (0..degree_vel)
```
Comment (james7132) [inline crates/bevy_math/src/bezier.rs:365]: Curious if the compiler will unroll this properly and elide bounds checking or not. Not worth blocking on but definitely something worth looking into.

File: benches/benches/bevy_math/bezier.rs [hunk 20]
```diff
@@ -0,0 +1,95 @@
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
```
Comment (james7132) [inline benches/benches/bevy_math/bezier.rs:1]: Thanks for adding proper benchmarks for this! Really helpful to have better perf checks for our more complex math operations (outside of what glam already provides).

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_math/src/bezier.rs (hunk 16 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:123]: I tried this initially but ran into some issues. I'll attempt again.

File: crates/bevy_math/src/bezier.rs [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+        // I tried pre-solving and caching 11 values along the curve at struct instantiation in an
+        // attempt to give the solver a better starting guess. This ended up being slightly slower,
+        // possibly due to the increased size of the type. Another option would be to store the last
+        // `t`, and use that, however it's possible this could end up in a bad state where t is very
+        // far from the naive but generally safe guess of x, e.g. after an animation resets.
+        //
+        // Further optimization might not be needed however - benchmarks are showing it takes about
+        // 50 nanoseconds for an ease operation on my modern laptop, which seems sufficiently fast.
+        let mut t_guess = x;
+        (0..Self::MAX_ITERS).any(|_| {
```
Comment (JoJoJet) [inline crates/bevy_math/src/bezier.rs:287]: This is a pretty surprising way of doing control flow. I feel it would be a lot easier to understand if this was rewritten to use `for _ in 0..SELF::MAX_ITERS`, and then using `break;` instead of returning `true`

File: crates/bevy_math/src/bezier.rs (hunk 16 above)
Comment (JoJoJet) [reply crates/bevy_math/src/bezier.rs:123]: You'll have to make sure the returned iterator has the lifetime of `self`. It will work if the return type is `impl Iterator<Item = P> + '_`

**Review by JoJoJet — CHANGES_REQUESTED** [review]
Nice documentation, it's extensive enough where necessary. The code all looks correct, but I'm quite concerned about the lack of tests. Earlier in the thread you gave two suggestions for ways of testing this. I think both of them would be good.

File: crates/bevy_math/src/bezier.rs (hunk 21 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:287]: The reason for this was to improve the chances of loop unrolling. Iterators are more likely to be properly unrolled, and the limited benchmarking I did suggested the iterator version was slightly faster. I spent quite a while attempting to optimize this function.

File: benches/benches/bevy_math/bezier.rs (hunk 20 above)
Comment (aevyrie) [reply benches/benches/bevy_math/bezier.rs:1]: I spent a while attempting to optimize the Bezier functions, so I needed some proper benchmarks. 😄

File: crates/bevy_math/src/bezier.rs (hunk 15 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:25]: Could you explain a bit more? I don't follow.

This trait is only used to reduce code repetition for implementing Beziers in multiple dimensions/types.

File: crates/bevy_math/src/bezier.rs (hunk 16 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:123]: I'm pretty sure I ended up with that signature, but now I don't recall why I moved away from it. 😥

File: crates/bevy_math/src/bezier.rs (hunk 19 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:365]: This isn't substitute for looking at the compiled ASM, but when I was benchmarking and optimizing, I found that these iterators were faster than equivalent for loops. I assumed it was for this reason.

Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:367]: Benchmarking building a cubic Bezier with 100 points, where those points are acceleration or position:

```
build_pos_cubic_100_points
                        time:   [384.77 ns 385.17 ns 385.60 ns]
Found 15 outliers among 100 measurements (15.00%)
  8 (8.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe

build_accel_cubic_100_points
                        time:   [384.53 ns 386.51 ns 388.01 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
```

File: crates/bevy_math/src/bezier.rs (hunk 16 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs:123]: Got it working, thanks for bringing this up!

**Comment by aevyrie** [comment]
Added tests and did some further cleanup.

**Review by JoJoJet — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@aevyrie once you fix CI I'll merge this :)

File: crates/bevy_math/src/bezier.rs (hunk 21 above)
Comment (aevyrie) [reply crates/bevy_math/src/bezier.rs]: I revisted benchmarks, and didn't find a difference, so I reverted back to a simpler for loop
